test/terraform_ssh_example_test.go

```diff
@@ -211,3 +220,196 @@ func testSCPToPublicHost(t *testing.T, terraformOptions *terraform.Options, keyP
+	// programatically emulate within this test. For both instances, we are using the Key Pair we created earlier,
+	// and the user "ubuntu", as we know the Instances are running an Ubuntu AMI that has such a user
+	publicHost := ssh.Host{
+		Hostname:    publicInstanceIP,
+		SshUserName: "ubuntu",
+		SshAgent:    true,
+	}
+	privateHost := ssh.Host{
+		Hostname:    privateInstanceIP,
+		SshUserName: "ubuntu",
+		SshAgent:    true,
+	}
+
+	// It can take a minute or so for the Instance to boot up, so retry a few times
+	maxRetries := 30
+	timeBetweenRetries := 5 * time.Second
+	description := fmt.Sprintf("SSH with Agent to private host %s via public host %s", publicInstanceIP, privateInstanceIP)
+
+	// Run a simple echo command on the server
+	expectedText := "Hello, World"
+	command := fmt.Sprintf("echo -n '%s'", expectedText)
+
+	// Instantiate a temporary SSH agent
+	socketDir, err := ioutil.TempDir("", "ssh-agent-")
+	if err != nil {
+		t.Fatal(err)
+	}
+	socketFile := filepath.Join(socketDir, "ssh_auth.sock")
+	os.Setenv("SSH_AUTH_SOCK", socketFile)
+	sshAgent := NewSSHAgent(socketDir, socketFile)
+	defer sshAgent.Stop()
+
+	// Create SSH key for the agent using the existing AWS SSH key pair
+	block, _ := pem.Decode([]byte(keyPair.KeyPair.PrivateKey))
+	pkey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		t.Fatal(err)
+	}
+	key := agent.AddedKey{PrivateKey: pkey}
+
+	// Verify that we can SSH to the Instance and run commands
+	retry.DoWithRetry(t, description, maxRetries, timeBetweenRetries, func() (string, error) {
+
+		// Add SSH key to the agent
+		keys, err := sshAgent.agent.List()
+		if len(keys) == 0 {
+			sshAgent.agent.Add(key)
+		}
+
+		actualText, err := ssh.CheckPrivateSshConnectionE(t, publicHost, privateHost, command)
+
+		if err != nil {
+			return "", err
+		}
+
+		if strings.TrimSpace(actualText) != expectedText {
+			return "", fmt.Errorf("Expected SSH command to return '%s' but got '%s'", expectedText, actualText)
+		}
+
+		return "", nil
+	})
+}
+
+type SSHAgent struct {
+	stop       chan bool
+	stopped    chan bool
+	socketDir  string
+	socketFile string
+	agent      agent.Agent
+	ln         net.Listener
+}
+
+// Create SSH agent, start it in background and returns control back to the main thread
+func NewSSHAgent(socketDir string, socketFile string) *SSHAgent {
+	s := &SSHAgent{make(chan bool), make(chan bool), socketDir, socketFile, agent.NewKeyring(), nil}
+	go s.run()
+	return s
+}
+
+// SSH Agent listner and handler
+func (s *SSHAgent) run() {
+	defer close(s.stopped)
+	var err error
+	s.ln, err = net.Listen("unix", s.socketFile)
+	if err != nil {
+		fmt.Errorf("Could not create socket %v", err)
+	}
+	for {
+		select {
+		case <-s.stop:
+			return
+		default:
+			c, err := s.ln.Accept()
+			if err != nil {
+				select {
+				case <-s.stop:
+					return
+				default:
+					fmt.Errorf("Could not accept connection to agent %v", err)
+				}
+				continue
```

Comment:
Why `continue`? If `err` is not `nil`, shouldn't you always return here?

Reply:
I guess one request process may fail but you don't want to shut the whole service down. I've moved the `continue` next to the error print plus a comment. If for some reason one request to the agent fails, `DoWithRetry` will handle the retry.